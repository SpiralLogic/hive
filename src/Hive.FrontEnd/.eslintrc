{
  "plugins": [
    "@typescript-eslint",
    "react-hooks",
    "testing-library",
    "jest-dom",
    "unicorn"
  ],
  "env": {
    "browser": true,
    "es6": true
  },
  "extends": [
    "kentcdodds",
    "plugin:jest/recommended",
    "plugin:jest-dom/recommended",
    "plugin:prettier/recommended",
    "plugin:testing-library/react",
    "plugin:testing-library/dom",
    "prettier",
    "plugin:unicorn/recommended",
    "preact"
  ],
  "rules": {
    "unicorn/no-null": "off",
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": [
      "error"
    ]
  },
  "globals": {
    "Atomics": "readonly",
    "SharedArrayBuffer": "readonly",
    "h": true
  },
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "./tsconfig.json",
    "jsx": true
  },
  "overrides": [
    {
      "files": [
        "**/*.jsx",
        "**/*.tsx"
      ],
      "rules": {
        "unicorn/filename-case": "off"
      }
    }
  ],
  "settings": {
    "react": {
      "pragma": "h"
    },
    "propWrapperFunctions": [
      "forbidExtraProps",
      {
        "property": "freeze",
        "object": "Object"
      }
    ],
    "linkComponents": [
      "Hyperlink",
      {
        "name": "Link",
        "linkAttribute": "to"
      }
    ]
  },
  "ignorePatterns": [
    "/node_modules/*/**",
    "/*.config.js",
    "/src/public/*/**",
    "**/__coverage__/**",
    "/src/index.tsx",
    "/src/index.html",
    "/build.js",
    "/inject.shim.js"
  ]
}
