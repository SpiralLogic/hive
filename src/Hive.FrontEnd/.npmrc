node-version=20.11.1