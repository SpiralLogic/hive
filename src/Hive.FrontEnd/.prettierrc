{
  "singleQuote": true,
  "printWidth": 110,
  "proseWrap": "always",
  "bracketSameLine": true,
  "endOfLine": "auto",
  "trailingComma": "es5",
  "tabWidth": 2
}
