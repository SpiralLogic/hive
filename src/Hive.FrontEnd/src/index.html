<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="description" content="Hive online board game">
  <title>Hive</title>
  [css]
  [scripts]    
</head>
</html>
