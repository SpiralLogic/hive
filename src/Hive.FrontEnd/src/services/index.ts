export * from './hextille-builder';
export * from './hive-dispatcher';
export * from './server-connection';
