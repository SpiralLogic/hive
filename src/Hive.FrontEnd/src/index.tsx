import { h, render } from 'preact';
import App from './components/App';
import { serverConnectionFactory } from './services';
import creatures from './svg/creatures.svg?raw';

import GameEngine from './services/game-engine';

document.body.insertAdjacentHTML('beforeend', `${creatures}`);
document.title = import.meta.env.HIVE_APP_TITLE;

const [, , gameId, currentPlayer] = window.location.pathname.split('/');
const engine = new GameEngine({ gameId, currentPlayer });
const properties = {
  engine,
  connectionFactory: serverConnectionFactory,
};

render(h(App, properties), document.body);

export { removeOtherPlayerMoves } from './utilities/hextille';
export { cellKey } from './utilities/hextille';
console.log(import.meta.env);
