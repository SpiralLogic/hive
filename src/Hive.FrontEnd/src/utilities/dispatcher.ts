import { useEffect } from 'preact/hooks';
import { HiveDispatcher, HiveEventListener, HiveIntent } from '../services';

const hiveDispatcher = new HiveDispatcher();

export const useHiveDispatcher = (): HiveDispatcher => hiveDispatcher;

export const addHiveDispatchListener = <T extends HiveIntent>(
  type: T['type'],
  listener: HiveEventListener<T>
): void => {
  useEffect(() => {
    const dispatcher = useHiveDispatcher();
    return dispatcher.add<T>(type, listener);
  });
};

export const dispatchHiveEvent = hiveDispatcher.dispatch;
