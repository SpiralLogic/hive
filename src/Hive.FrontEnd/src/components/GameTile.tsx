import { FunctionComponent } from 'preact';
import { useEffect, useState } from 'preact/hooks';

import { PlayerId, Tile as TileType } from '../domain';
import { TileAction } from '../services';
import { useClassReducer } from '../utilities/class-reducer';
import { useHiveDispatchListener, dispatchHiveEvent } from '../utilities/dispatcher';
import { handleDrop, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import Tile from './Tile';

const tileSelector = `[tabindex].tile`;
const cellSelector = `[tabindex][role="cell"]`;
type Properties = TileType & { stacked?: boolean; currentPlayer: PlayerId };

const handleMouseLeave = (event: { currentTarget: HTMLElement }) => event.currentTarget.blur();

const GameTile: FunctionComponent<Properties> = (properties) => {
  const { currentPlayer, stacked, ...tile } = properties;
  const { id, moves, creature, playerId } = tile;
  const [focus, setFocus] = useState(tileSelector);
  const [classes, setClassList] = useClassReducer(`player${playerId} hex`);
  if (stacked) setClassList({ type: 'add', classes: ['stacked'] });

  const deselect = (fromEvent = false) => {
    if (!classes.includes('selected')) return;
    setClassList({ type: 'remove', classes: ['selected'] });
    dispatchHiveEvent({ type: 'tileDeselected', tile, fromEvent });
  };

  const select = (fromEvent = false) => {
    if (classes.includes('selected')) return;
    dispatchHiveEvent({ type: 'tileClear' });
    setClassList({ type: 'add', classes: ['selected'] });
    if (currentPlayer != playerId) return;
    dispatchHiveEvent({ type: 'tileSelected', tile, fromEvent });
  };

  useHiveDispatchListener<TileAction>('tileSelect', (event: TileAction) => {
    if (event.tile.id === id) select(true);
  });

  useHiveDispatchListener<TileAction>('tileDeselect', (event: TileAction) => {
    if (event.tile.id === id) deselect(true);
  });

  useHiveDispatchListener('tileClear', () => {
    deselect(true);
    setFocus('');
  });

  const handleDragStart = (event: DragEvent) => {
    event.stopPropagation();
    select();
    setClassList({ type: 'add', classes: ['before-drag'] });
    setTimeout(() => setClassList({ type: 'remove', classes: ['before-drag'] }), 1);
  };

  const handleDragEnd = () => {
    dispatchHiveEvent({ type: 'tileClear' });
    dispatchHiveEvent({ type: 'tileDropped', tile });
  };

  const handleClick = (event: MouseEvent) => {
    event.stopPropagation();
    return classes.includes('selected') ? deselect() : select();
  };

  const handleKeyDown = (event: KeyboardEvent) => {
    if (handleKeyboardNav(event) || !isEnterOrSpace(event)) return;
    event.stopPropagation();
    if (classes.includes('selected')) {
      deselect();
    } else {
      select();
      setFocus(cellSelector);
    }
  };
  useEffect(() => {
    if (!focus) return;
    const focusElement = document.querySelector<HTMLElement>(focus);
    focusElement?.focus();
    setFocus('');
  }, [focus]);

  const attributes = {
    title: `Player-${playerId} ${creature}`,
    class: classes,
    draggable: moves.length > 0 ? true : undefined,
    tabindex: moves.length > 0 ? 0 : undefined,
    creature,
  };

  const handlers =
    moves.length > 0
      ? {
          onclick: handleClick,
          ondragstart: handleDragStart,
          ondragend: handleDragEnd,
          onkeydown: handleKeyDown,
          onmouseleave: handleMouseLeave,
          ondrop: handleDrop,
        }
      : {
          ondrop: handleDrop,
        };
  return <Tile {...attributes} {...handlers} />;
};
GameTile.displayName = 'GameTile';
export default GameTile;
