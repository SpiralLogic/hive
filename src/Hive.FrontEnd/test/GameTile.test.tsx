import { act, fireEvent, screen, render } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { h } from 'preact';
import { HiveEvent, TileAction, TileEvent } from '../src/services';
import { useHiveDispatcher } from '../src/utilities/dispatcher';
import GameTile from '../src/components/GameTile';
import { simulateEvent } from './test-helpers';

describe('tile Tests', () => {
  const tileCanMove = {
    id: 1,
    playerId: 1,
    creature: 'tileCanMove',
    moves: [{ q: 1, r: 1 }],
  };
  const tileNoMove = { id: 2, playerId: 0, creature: 'tileNoMove', moves: [] };

  const createTileCanMove = () => {
    render(<GameTile currentPlayer={1} {...tileCanMove} />);
    return screen.getByTitle(/tileCanMove/);
  };

  const createTileNoMove = () => {
    render(<GameTile currentPlayer={1} {...tileNoMove} />);
    return screen.getByTitle(/tileNoMove/);
  };

  const createDispatcher = (): [TileEvent[], () => void] => {
    const tileEvents: TileEvent[] = [];
    const listener = (e: TileEvent) => tileEvents.push(e);

    const dispatcher = useHiveDispatcher();
    const cleanup = dispatcher.add<TileEvent>('tileSelected', listener);
    return [tileEvents, cleanup];
  };

  describe('tile render', () => {
    it('has creature', () => {
      const tileCanMoveElement = createTileCanMove();
      const tileNoMoveElement = createTileNoMove();
      expect(tileCanMoveElement).toMatchSnapshot();
      expect(tileNoMoveElement).toMatchSnapshot();
    });
  });

  describe('tile events', () => {
    it('click emits tile start event', () => {
      const [tileEvents, cleanup] = createDispatcher();
      const tileCanMoveElement = createTileCanMove();
      userEvent.click(tileCanMoveElement);
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(tileEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
      cleanup();
    });

    it('enter emits tile start event', () => {
      const [tileEvents, cleanup] = createDispatcher();
      const tileCanMoveElement = createTileCanMove();
      userEvent.type(tileCanMoveElement, '{enter}');
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
        fromEvent: false,
      };
      expect(tileEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
      cleanup();
    });

    it('enter selects tile', async () => {
      const tileCanMoveElement = createTileCanMove();
      tileCanMoveElement.focus();
      userEvent.keyboard('{enter}');

      expect(await screen.findByTitle(/tileCanMove/)).toHaveClass('selected');
    });

    it('second enter deselects tile', async () => {
      const tileCanMoveElement = createTileCanMove();

      tileCanMoveElement.focus();
      userEvent.keyboard('{enter}');
      expect(await screen.findByTitle(/tilecanmove/i)).toHaveClass('selected');
      userEvent.keyboard('{enter}');

      expect(await screen.findByTitle(/tilecanmove/i)).not.toHaveClass('selected');
    });

    it('arrow keys use handler', () => {
      const tileCanMoveElement = createTileCanMove();

      tileCanMoveElement.focus();
      userEvent.keyboard('{arrowdown}');

      expect(tileCanMoveElement).toHaveFocus();
    });

    it('space emits tile start event', () => {
      const [tileEvents, cleanup] = createDispatcher();
      const tileCanMoveElement = createTileCanMove();

      tileCanMoveElement.focus();
      userEvent.type(tileCanMoveElement, ' ');
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(tileEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
      cleanup();
    });

    it('click selects previous selected tile', async () => {
      const tileCanMoveElement = createTileCanMove();

      userEvent.click(tileCanMoveElement);
      expect(await screen.findByTitle(/tileCanMove/)).toHaveClass('selected');
      userEvent.click(tileCanMoveElement);

      expect(await screen.findByTitle(/tileCanMove/)).not.toHaveClass('selected');
    });
    it('click deselects previous selected tile', async () => {
      const tileCanMoveElement = createTileCanMove();

      userEvent.click(tileCanMoveElement);

      expect(await screen.findByTitle(/tileCanMove/)).toHaveClass('selected');
    });

    it('clicking same tile doesnt fire a tile start event', () => {
      const mock = jest.spyOn(useHiveDispatcher(), 'dispatch');
      const tileCanMoveElement = createTileCanMove();

      userEvent.click(tileCanMoveElement);

      mock.mockClear();
      userEvent.click(tileCanMoveElement);

      expect(useHiveDispatcher().dispatch).not.toHaveBeenCalledWith(
        expect.objectContaining({ type: 'tileSelect' })
      );
    });

    it('mouseLeave activates blur', () => {
      const tileCanMoveElement = createTileCanMove();
      tileCanMoveElement.focus();
      userEvent.unhover(tileCanMoveElement);
      expect(tileCanMoveElement).not.toHaveFocus();
    });

    it('is draggable when there are available moves', () => {
      const tileCanMoveElement = createTileCanMove();
      expect(tileCanMoveElement).toHaveAttribute('draggable', 'true');
    });

    it('is *not* draggable when there are no moves available', () => {
      const tileNoMoveElement = createTileNoMove();
      expect(tileNoMoveElement).not.toHaveAttribute('draggable', 'false');
    });

    it('on drag emits start event', () => {
      const [tileEvents, cleanup] = createDispatcher();
      const tileCanMoveElement = createTileCanMove();
      fireEvent.dragStart(tileCanMoveElement);
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(tileEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
      cleanup();
    });

    it('on dragEnd emits end event', () => {
      const dropEvents: TileEvent[] = [];
      const tileCanMoveElement = createTileCanMove();
      useHiveDispatcher().add<TileEvent>('tileDropped', (e) => dropEvents.push(e));
      fireEvent.dragEnd(tileCanMoveElement);
      const expectedEvent: HiveEvent = {
        type: 'tileDropped',
        tile: tileCanMove,
      };

      expect(dropEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
    });

    it(`tile can be selected via action`, () => {
      const tileCanMoveElement = createTileCanMove();

      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      }).catch(() => {});
      expect(tileCanMoveElement).toHaveClass('selected');
    });

    it(`an already selected tile doesnt fire a selected event when selected`, () => {
      const selectEvents: TileEvent[] = [];
      useHiveDispatcher().add<TileEvent>('tileSelected', (e) => selectEvents.push(e));
      createTileCanMove();
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      }).catch(() => {});
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      }).catch(() => {});
      expect(selectEvents).toHaveLength(1);
    });

    it(`an already deselected tile doesnt fire a deselected event when deselected`, () => {
      const deselectEvents: TileEvent[] = [];
      useHiveDispatcher().add<TileEvent>('tileDeselected', (e) => deselectEvents.push(e));
      createTileCanMove();
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      }).catch(() => {});
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileDeselect', tile: tileCanMove });
      }).catch(() => {});
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileDeselect', tile: tileCanMove });
      }).catch(() => {});
      expect(deselectEvents).toHaveLength(1);
    });

    it(`tile can be deselected via action`, async () => {
      const tileCanMoveElement = createTileCanMove();

      await act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      });
      await act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileDeselect', tile: tileCanMove });
      });
      expect(tileCanMoveElement).not.toHaveClass('selected');
    });

    it(`tile is only selected on matching select events`, () => {
      const tileNoMoveElement = createTileNoMove();
      useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileNoMove });

      expect(tileNoMoveElement).not.toHaveClass('selected');
    });

    it(`tile is only deselected on matching select events`, () => {
      createTileCanMove();
      createTileNoMove();
      const deselectEvents: TileEvent[] = [];
      useHiveDispatcher().add<TileEvent>('tileDeselected', (e) => deselectEvents.push(e));
      useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      useHiveDispatcher().dispatch<TileAction>({ type: 'tileDeselect', tile: tileNoMove });

      expect(deselectEvents).toHaveLength(0);
    });

    it('default on drop is prevented', () => {
      const tileCanMoveElement = createTileCanMove();
      const tileNoMoveElement = createTileNoMove();
      expect(simulateEvent(tileCanMoveElement, 'drop')).toHaveBeenCalledWith();
      expect(simulateEvent(tileNoMoveElement, 'drop')).toHaveBeenCalledWith();
    });
  });

  describe('tile Snapshot', () => {
    it('can move matches current snapshot', () => {
      const tileCanMoveElement = createTileCanMove();

      expect(tileCanMoveElement).toMatchSnapshot();
    });

    it('no moves matches current snapshot', () => {
      const tileNoMoveElement = createTileNoMove();
      expect(tileNoMoveElement).toMatchSnapshot();
    });

    it('when stacked matched snapshot', () => {
      render(<GameTile currentPlayer={0} {...tileNoMove} stacked={true} />);
      expect(screen.getByTitle(/Player-0/)).toMatchSnapshot();
    });
  });
});
