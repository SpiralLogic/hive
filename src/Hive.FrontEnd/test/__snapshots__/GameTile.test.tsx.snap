// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tile Tests tile Snapshot can move matches current snapshot 1`] = `
<div
  class="player1 hex tile tilecanmove"
  draggable="true"
  tabindex="0"
  title="Player-1 tileCanMove"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature"
      href="#tilecanmove"
    />
  </svg>
</div>
`;

exports[`tile Tests tile Snapshot no moves matches current snapshot 1`] = `
<div
  class="player0 hex tile tilenomove"
  title="Player-0 tileNoMove"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature"
      href="#tilenomove"
    />
  </svg>
</div>
`;

exports[`tile Tests tile Snapshot when stacked matched snapshot 1`] = `
<div
  class="player0 hex stacked tile tilenomove"
  title="Player-0 tileNoMove"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature"
      href="#tilenomove"
    />
  </svg>
</div>
`;

exports[`tile Tests tile render has creature 1`] = `
<div
  class="player1 hex tile tilecanmove"
  draggable="true"
  tabindex="0"
  title="Player-1 tileCanMove"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature"
      href="#tilecanmove"
    />
  </svg>
</div>
`;

exports[`tile Tests tile render has creature 2`] = `
<div
  class="player0 hex tile tilenomove"
  title="Player-0 tileNoMove"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature"
      href="#tilenomove"
    />
  </svg>
</div>
`;
